### ISSUE TITLE
[Bug] 'LDAP operation [rename] failed.' OpenLdap

### ISSUE THREAD

**Issue by nohnaimer** [DirectoryTree/LdapRecord#253]
**Environment (please complete the following information):**
 - LDAP Server Type: OpenLDAP
 - PHP Version: 7.4.16
 - LdapRecord 2.2.3

**Describe the bug:**
```bash
Exception 'Exception' with message 'LDAP operation [rename] failed.'

in /Users/and1/phpProjects/www/ems/vendor/directorytree/ldaprecord/src/LdapBase.php:168

Stack trace:
#0 /Users/and1/phpProjects/www/ems/vendor/directorytree/ldaprecord/src/Ldap.php(307): LdapRecord\LdapBase->executeFailableOperation(Object(Closure))
#1 /Users/and1/phpProjects/www/ems/vendor/directorytree/ldaprecord/src/Query/Builder.php(1629): LdapRecord\Ldap->rename('cn=\xD0\x9F\xD1\x80\xD0\xB0\xD0\xB4\xD1\x83\xD0\xBD...', 'cn=\\d0\\9b\\d1\\8f...', 'ou=\xD0\x93\xD0\xBE\xD0\xBC\xD0\xB5\xD0\xBB\xD1\x8C...', true)
#2 /Users/and1/phpProjects/www/ems/vendor/directorytree/ldaprecord/src/Connection.php(397): LdapRecord\Query\Builder->LdapRecord\Query\{closure}(Object(LdapRecord\Ldap))
#3 /Users/and1/phpProjects/www/ems/vendor/directorytree/ldaprecord/src/Connection.php(355): LdapRecord\Connection->runOperationCallback(Object(Closure))
#4 /Users/and1/phpProjects/www/ems/vendor/directorytree/ldaprecord/src/Query/Builder.php(1630): LdapRecord\Connection->run(Object(Closure))
#5 /Users/and1/phpProjects/www/ems/vendor/directorytree/ldaprecord/src/Models/Model.php(1263): LdapRecord\Query\Builder->rename('cn=\xD0\x9F\xD1\x80\xD0\xB0\xD0\xB4\xD1\x83\xD0\xBD...', 'cn=\\d0\\9b\\d1\\8f...', 'ou=\xD0\x93\xD0\xBE\xD0\xBC\xD0\xB5\xD0\xBB\xD1\x8C...', true)
#6 /Users/and1/phpProjects/www/ems/common/models/behaviors/AddressBookEmployeeBehavior.php(178): LdapRecord\Models\Model->rename('cn=\\d0\\9b\\d1\\8f...')
#7 /Users/and1/phpProjects/www/ems/common/models/behaviors/AddressBookEmployeeBehavior.php(120): common\models\behaviors\AddressBookEmployeeBehavior->update()
#8 [internal function]: common\models\behaviors\AddressBookEmployeeBehavior->afterUpdate(Object(yii\db\AfterSaveEvent))
#9 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/base/Component.php(628): call_user_func(Array, Object(yii\db\AfterSaveEvent))
#10 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/db/BaseActiveRecord.php(1000): yii\base\Component->trigger('afterUpdate', Object(yii\db\AfterSaveEvent))
#11 /Users/and1/phpProjects/www/ems/common/models/Employee.php(341): yii\db\BaseActiveRecord->afterSave(false, Array)
#12 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/db/BaseActiveRecord.php(825): common\models\Employee->afterSave(false, Array)
#13 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/db/ActiveRecord.php(681): yii\db\BaseActiveRecord->updateInternal(NULL)
#14 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/db/BaseActiveRecord.php(681): yii\db\ActiveRecord->update(true, NULL)
#15 /Users/and1/phpProjects/www/ems/console/models/SyncProsellerModel.php(69): yii\db\BaseActiveRecord->save()
#16 /Users/and1/phpProjects/www/ems/console/controllers/EmployeeController.php(155): console\models\SyncProsellerModel->sync('/Users/and1/Des...')
#17 [internal function]: console\controllers\EmployeeController->actionSync()
#18 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#19 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/base/Controller.php(181): yii\base\InlineAction->runWithParams(Array)
#20 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/console/Controller.php(184): yii\base\Controller->runAction('sync', Array)
#21 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/base/Module.php(534): yii\console\Controller->runAction('sync', Array)
#22 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/console/Application.php(181): yii\base\Module->runAction('employee/sync', Array)
#23 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/console/Application.php(148): yii\console\Application->runAction('employee/sync', Array)
#24 /Users/and1/phpProjects/www/ems/vendor/yiisoft/yii2/base/Application.php(392): yii\console\Application->handleRequest(Object(yii\console\Request))
#25 /Users/and1/phpProjects/www/ems/yii(27): yii\base\Application->run()
#26 {main}
```
Cn in OpenLdap and rdn in code - different. Try admin user - same error.
I use same method rename https://github.com/DirectoryTree/LdapRecord/blob/c7221d46b64e5a301a4820e583190a79c50c8013/src/Models/Model.php#L1239 described https://ldaprecord.com/docs/core/v2/model-api/#rename
Bug fix - https://github.com/DirectoryTree/LdapRecord/issues/237
Any ideas?

**Comment by stevebauman** [comment]
Hi @nohnaimer, can you post your code with the values that you're setting for the objects new RDN?

Make sure you provide the objects new RDN into the rename method:

```php
$user = User::find('cn=User,dc=local,dc=com');

$user->rename('cn=Renamed User');
```

**Comment by nohnaimer** [comment]
Code:
```php
        $user = User::query()->findBy('EmployeeID', 7150);
        $user->rename('cn=Jane Doe');
        $user->refresh();
```
On $user I get user object with data from openldap server, but I still get error LDAP operation [rename] failed.
<img width="2472" alt="Снимок экрана 2021-03-16 в 17 15 19" src="https://user-images.githubusercontent.com/8865830/111323744-36989200-867b-11eb-9a93-1c0fbc1325d8.png">

**Comment by stevebauman** [comment]
Hmm okay we'll have to pull the detailed error to see more here.

Wrap the `rename` call in the `try/catch` and then dump the detailed error object:

```php
try {
     $user->rename('cn=Jane Doe');
} catch (\Exception $e) {
    $ldap = $user->getConnection()->getLdapConnection();

    dd($ldap->getDetailedError());
}
```

Then, post the details of the object here 👍

**Comment by nohnaimer** [comment]
LdapRecord\DetailedError Object
(
    [errorCode:protected] => 34
    [errorMessage:protected] => Invalid DN syntax
    [diagnosticMessage:protected] => invalid newSuperior
)

**Comment by stevebauman** [comment]
> invalid newSuperior

Hmm, this error means that the OU that the object resides in is invalid (at least looking at the OpenLDAP documentation).

Does renaming any other object work in your directory? Can you confirm that?

**Comment by nohnaimer** [comment]
@stevebauman Hi,
You du not encode dn by rename and do not like my DN cn=Jane Doe,ou=Гродно\2C ул.Красноармейская\2C 3,l=Гродно,o=ProSeller,ou=Personnel,dc=test,dc=loc or when create DN do something wrong.

My escape method:
```php
/**
     * @param string $value
     * @return string
     */
    public function escape($value): string
    {
        return addcslashes(addslashes($value), ',');
    }
```
Not escaped value = Гродно, ул.Красноармейская, 3

I do not exclude that I am doing something wrong.

**Comment by nohnaimer** [comment]
Try to change my escape method:
```php
/**
     * @param string $value
     * @return string
     */
    public function escape($value): string
    {
        return ldap_escape($value);
//        return addcslashes(addslashes($value), ',');
    }
```
Same escaped DN - cn=Jane Doe,ou=Барановичи\2C пл.Ленина\2C1,l=Брест,o=ProSeller,ou=Personnel,dc=test,dc=loc and same error.
Any ideas how to escape('Барановичи, пл.Ленина,1') ?

**Comment by stevebauman** [comment]
Hey @nohnaimer -- I haven't forgotten about you and this issue. I'll have a patch out shortly that resolves this problem! 👍

**Comment by stevebauman** [comment]
Hey @nohnaimer! I've just pushed a patch to resolve this -- could you try using `dev-master` and attempt renaming using the below code?

```php
$user = User::query()->findBy('EmployeeID', 7150);

$user->rename(
    $user->getCreatableRdn('Барановичи, пл.Ленина,1')
);

$user->refresh();
```

**Comment by nohnaimer** [comment]
```php
/** @var User $user */
        $user = User::query()->findBy('EmployeeID', 7164);
        try {
            $user->rename($user->getCreatableRdn('Jane Doe', 'cn'));
        } catch (\Exception $e) {
            $ldap = $user->getConnection()->getLdapConnection();
            debug($ldap->getDetailedError());
        }
```
Same error because 'Барановичи, пл.Ленина,1' part of DN. I change cn (rename) in this case. Do not escape DN in method rename.

**Comment by stevebauman** [comment]
That's strange -- can you dump all of your users attributes using `var_dump($user->getAttributes())` and then post the result here so I can create a test case with all of the properties? Namely, the raw `dn` and `cn` attributes is what I'm looking for. Pease remove any sensitive information before posting, of course.

Once I have that I'll be able to debug this locally a lot easier. 👍

**Comment by nohnaimer** [comment]
HI @stevebauman.
On method getAttributes() no dn information from OpenLdap. I debug data on this. code:
```php
$user = User::query()->findBy('EmployeeID', 7164);
debug($user->getAttributes(), $user->getDn());
```
Result:
```php
Array
(
    [0] => Array
        (
            [cn] => Array
                (
                    [0] => Иванов Иван Иванович
                )

            [employeeid] => Array
                (
                    [0] => 7164
                )

            [l] => Array
                (
                    [0] => Брест
                )

            [o] => Array
                (
                    [0] => Company
                )

            [title] => Array
                (
                    [0] => Договорник
                )

            [userdepartmentid] => Array
                (
                    [0] => 720
                )

            [uid] => Array
                (
                    [0] => ivanov
                )

            [mail] => Array
                (
                    [0] => ivanov@test.loc
                )

            [givenname] => Array
                (
                    [0] => Иван
                )

            [sn] => Array
                (
                    [0] => Иванов
                )

            [objectclass] => Array
                (
                    [0] => inetOrgPerson
                )

            [entryuuid] => Array
                (
                    [0] => 448c4272-2a1f-103b-96f1-3f5a40193fa8
                )

        )

    [1] => cn=Иванов Иван Иванович,ou=Барановичи\2C ул.Советская\2C84,l=Брест,o=Company,ou=Personnel,dc=test,dc=loc
)
```

**Comment by stevebauman** [comment]
Excellent, thanks @nohnaimer! I'm able to reproduce this issue locally. I'll have a bug fix for this out tomorrow evening 👍 

It's due to the `DistinguishedName` class un-escaping the distinguished name components, and producing an invalid DN.

**Comment by stevebauman** [comment]
Ok -- I've resolved this issue. Can you try updating to `dev-master` and attempt a rename with that account?

**Comment by nohnaimer** [comment]
Hi @stevebauman,
Everything works now.
Please add git tag because dev-master (2.4.x-dev) do not download. In composer.json lacks:
```bash
"extra": {
        "branch-alias": {
            "dev-master": "2.4.x-dev"
        }
    }
```

**Comment by stevebauman** [comment]
Awesome! That's great to hear @nohnaimer. I'm glad you're up and running.

And absolutely, I've just released [v2.4.2](https://github.com/DirectoryTree/LdapRecord/releases/tag/v2.4.2) that contains this patch 👍
